src/devices/cpu/psx/dma.cpp

```diff
@@ -242,7 +242,20 @@ void psxdma_device::write(offs_t offset, uint32_t data, uint32_t mem_mask)
 				{
 					LOG( "%s: dma %d read block %08x %08x\n", machine().describe_context(), index, n_address, n_size );
 					dma->fn_read( m_ram, n_address, n_size );
-					dma_finished( index );
+
+					if( index == 5 )
+					{
+						// Required for namcos12 CDXA-based games
+						// These games clears the internal DMA status flag after starting the DMA and expects the DMA finished
+						// callback, which normally should be called sometime after the init code has finished executing but is
+						// instead called immediately without using dma_start_timer, to set the status flag to a finished state
+						// while it loops indefinitely waiting for the DMA status flag to change.
+						dma_start_timer( index, 500 );
+					}
+					else
+					{
+						dma_finished( index );
+					}
```

Comment:
It's a workaround for a bigger issue. Channel 5 itself isn't special but I wanted to limit the scope of the workaround to what I could test.

Reply:
This is a *really* nasty kind of hack.  Putting a hack like this into a device to work around an issue is liable to break other software.  I’d be really uncomfortable with this given it affects everything with a PlayStation chipset.  These kinds of hacks also have a habit of sticking around a lot longer than they should.

Reply:
It's either per-game hacks for all games that have broken DMAs or adding in a simulated delay for DMAs like what's in this PR. Arguably this is a better solution than per-game hacks. DMAs ending instantly wasn't ever a sane default. I'm pretty sure I've mentioned it before but the PSX DMA device needs a full rewrite, and it's not coming from me.

If you'd prefer to keep the broken DMAs and for me to add in per-game hacks for truckk and ujlnow then I can do that.

Reply:
Can it do not-instant DMA on all channels rather than hacking just channel 5?  Should the delay be affected by the amount of data transferred? 

A magic number applied to just one channel because it happens to make certain games work just doesn’t sit right. Whether you change it for one channel or all the channels, it still has the potential to affect everything with a PlayStation chipset.

Reply:
It could, but if you're worried about breakage then that's where you're going to really see potential breakage. I think having it be affected by the amount of data transferred is probably more correct but that's where I'm less confident. 

Just to give as much info as needed to understand the change: channel 5 specifically means PIO DMAs. You can't specify each DMA to use a specific channel defined on the user's/programmer's end (like, PIO DMAs will never be channel 2, channel 2 will always be GPU DMAs). They all have special meanings for the specific hardware they're meant to interface with. I don't know of any normal PS1 games that use PIO DMAs off the top of my head and the handful of PS1 arcade games I know of that use it are fixed by this change. 

I can change it to use the PIO DMA timings based on previous research (20 cycles per 0x100 words) if it'd help at all. I feel like any deeper changes would really require much more testing and I'd prefer to just revert this change and add in per-game hacks.

https://psx-spx.consoledev.net/dmachannels/#dma-transfer-rates

Reply:
The transfer rates on that page are best-case, aren’t they?  Calculating delays based on those rates should still be better than instantaneous while being unlikely to break anything due to being too slow.

Reply:
I guess the potential for issues even if optimistic timings are used comes from not emulating the DMA controller stalling the CPU.  It’s conceivably possible that some game expects fewer CPU cycles to execute while the DMA transfer is in progress because it blocks a cache fill, temporarily halting the CPU.

Reply:
Ok. I think in this case then it'd be better to revert this change and add in the per-game hacks. I know enough to be able to diagnose the DMA timing issue but also recognize that I don't think I can give the root issue an actual proper fix (at a quick glance multiple channels have configurable DMA timings so it'd be more than just dma.cpp that needs changes, plus anything with the CPU halting I just don't know about). I think anything more would really need to be its own PR and require a lot more testing because something *will* break. The DMA device already has some long standing hacks and notes that taking too long on some DMAs causes certain games to break, for example.

Reply:
Turns out the issue does not affect ujlnow, only ptblank2, gunbarl, and truckk. ujlnow's routine for using the PIO DMA sets its status flag before starting the DMA. One less hack required.